### ISSUE TITLE
Any possibility to export a C++ learned Net ?  

### ISSUE THREAD

**Issue by DerHaddy** [pytorch/pytorch#26086]
Hey,

does anyone know about a possibility of exporting and importing a learned model in C++?

I found the solution to use:

```
string model_path = "model.pt";
serialize::OutputArchive output_archive;
model.save(output_archive);
output_archive.save_to(model_path); 
```

But the outcome is a zip file with a .json header and just the weight and bias tensors as a file. When open with Netron, it just shows me the saved tensors without any connection. I tried to import the C++ exported model to Python to use there the ONNX export. But unfortunatly to that it needs the whole method of the forward pass, because it is not saved in the exported file. I would like to export a whole net in C++, including forward pass, netstructure ... like it is in the ONNX format.

Is there any solution to that kind of problem?

**Comment by DerHaddy** [comment]
Thanks for you reply,

if i see that right and to make it clear, there is no possibility to convert a normal PyTorch C++ net to TorchScript in C++ API? Because for me it is relevant to export a C++ trained model and to save it and open it in an extern C++ project where I need to be able to use the forward pass. And there are no different solutions to it? 

That would be pretty sad, because the C++ API is pretty good overall.

Thanks a lot.

**Comment by driazati** [comment]
Right, today there's no way to serialize the code of a C++ frontend-created model. The external project will have to have the library containing the model linked in so it can instantiate it itself (then you can load in the parameters).

**Comment by DerHaddy** [comment]
Is there a plan to implement a feature which supports this? Because in my opinion it is a pretty important feature but anyways, a big thanks for providing this information.
 
According to that a ONNX import or export would be great as well, but thats another thing.

**Comment by charlesrwest** [comment]
If I may ask, is there any way to expedite the creation of support for this feature (script export from C++ frontend)?  I'm not familiar enough with the codebase to do this in an at all reasonable amount of time, but I would happily through $50 in the pot if it would expedite things (grad student, otherwise would offer more).

I would really rather not have to have two separate pipelines for doing inference depending on the language.

**Comment by suo** [comment]
Thanks for the feedback! Let me share our thinking around this area.

We have no current plans to support script export from the C++ frontend. Doing so is a very high-complexity project (multiple full time engineers working on it basically indefinitely), and we don't think it's worth it due to the number of alternatives available. cc @yf225 who is the maintainer for this area.

Here are some of the alternatives:
- If you have a C++ model and want to use it in a different C++ project (@DerHaddy's use case), there are many solutions: link the model in as a shared library, compile it to a binary and feed it input from disk, etc. Calling C++ from C++ is solved(-ish) problem.
- If you have a C++ model and want to call it from Python so you have only one pipeline for doing inference (@charlesrwest's use case), you can implement it as a C++ extension to PyTorch. ([tutorial](https://pytorch.org/tutorials/advanced/cpp_extension.html)). 
- Of course, you can always implement the model in TorchScript and get C++/Python portability for free.

I understand that it's a weird inconsistency in our overall API surface though :(

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
